File: dart/angular/test/button_unit_spec.dart
import 'dart:async';
import 'dart:html';

import 'package:unittest/unittest.dart' hide expect;
import 'package:guinness/guinness.dart';

import 'package:angular/angular.dart';
import 'package:angular/mock/module.dart';

import 'package:angular_dart_examples/components/button/button_module.dart';

loadTemplates(List<String> templates) {
	return () {
		updateCache(template, response) => inject((TemplateCache cache) => cache.put(template, response));

		final futures = templates.map((template) => HttpRequest.request('../lib/$template', method: 'GET')
			.then((_) => updateCache('../lib/$template', new HttpResponse(200, _.response))));

		return Future.wait(futures);
	};
}

void main(){
	describe('ButtonComponent', () {
		beforeEach(setUpInjector);
		afterEach(tearDownInjector);

		beforeEach(module((Module module) => module..bind(ButtonComponent)));
		beforeEach(loadTemplates(['buttons-template.html', 'button-template.html']));

		it('should render', () {
			async(inject((TestBed testBed, Scope scope) {
				scope.context['info'] = {'key': 'value'};
				scope.context['procedure'] = (Event event) => window.console.log(event);

				final HtmlElement element = testBed.compile(
						'<ui-button name="button" version="outline" color="red"'
						+ 'procedure="procedure" info="info" class="custom">Button</ui-button>'
						, scope: scope);

				microLeap();
				testBed.rootScope.apply();

				final ShadowRoot root = element.shadowRoot;
				final button = root.querySelector('button');

				expect(button.text).toEqual('Button');

				expect(button.getAttribute('name')).toEqual('button');

				expect(button.getAttribute('class')).toContain('custom');
				expect(button.getAttribute('class')).toContain('button');
				expect(button.getAttribute('class')).toContain('button--outline');
				expect(button.getAttribute('class')).toContain('button--red');
			}));	
		});
	});
}

File: javascript/angular/components/button/test/button-unit-spec.js
describe('button directive', function() {

	beforeEach(module('ButtonModule'));

	// Define
	var compile,
		httpBackend,
		rootScope,
		scope,
		elementScope,
		controller,
		buttonController,
		template,
		component,
		element,
		button;

	// Assign
	beforeEach(inject(function($compile, $rootScope, $controller, $httpBackend) {
		compile = $compile;
		rootScope = $rootScope;
		controller = $controller;
		httpBackend = $httpBackend;
		scope = rootScope.$new();
		template =
			'<button ng-class="classes">'
				+ '<span ng-transclude></span>'
			+ '</button>';
		component =
			'<ui-button'
				+ ' name="button"'
				+ ' version="outline"'
				+ ' color="red"'
				+ ' procedure="click(\'click!\')"'
				+ ' info="object"'
				+ ' class="custom">'
						+ 'Button'
			+ '</ui-button>';
	}));

	// Initialize
	beforeEach(function() {
		scope.version = 'outline';
		scope.color = 'red';
		scope.object = {key: 'value'};
		scope.click = jasmine.createSpy('click');

		// buttonController = controller('ButtonController');
		httpBackend.expectGET('src/button-template.html').respond(template);
		element = compile(component)(scope);

		scope.$digest();
		httpBackend.flush();

		elementScope = element.isolateScope();
		button = element.find('button');
	});

	// Cleanup
	afterEach(function() {
		compile = null;
		rootScope = null;
		controller = null;
		buttonController = null;
		httpBackend = null;
		scope = null;
		elementScope = null;
		template = null;
		component = null;
		element = null;
		button = null;
	});

	// Scope
	it('should populate the scope', function() {
		expect(elementScope.version).toEqual(scope.version);
		expect(elementScope.color).toEqual(scope.color);
		expect(elementScope.info).toEqual(scope.object);
		expect(typeof(elementScope.procedure)).toEqual('function');

		elementScope.procedure();
		expect(scope.click).toHaveBeenCalledWith('click!');
	});

	// Controller
	it('should bind properties to controller', function() {
		expect(elementScope.button.version).toEqual(scope.version);
		expect(elementScope.button.color).toEqual(scope.color);
		expect(elementScope.button.info).toEqual(scope.object);
		expect(typeof(elementScope.button.procedure)).toEqual('function');
	});

	// Element
	it('should render html', function() {
		expect(button.text()).toEqual('Button');
		expect(button.attr('name')).toEqual('button');

		expect(button.attr('class')).toContain('custom');
		expect(button.attr('class')).toContain('button');
		expect(button.attr('class')).toContain('button--outline');
		expect(button.attr('class')).toContain('button--red');
	});

	// Events
	it('should handle events', function() {
		element.triggerHandler('click');
		expect(scope.click).toHaveBeenCalledWith('click!');
	});
});